Diag axis obj

uramirez8707 opened the pull request:
**Description**
This add a diag_axis object to be used in the modern diag_manager. 
There are still a lot of todos. 

**How Has This Been Tested?**
CI

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+    CHARACTER(len=*),   INTENT(in)           :: name            !< Name of the axis
+    class(*),           INTENT(in)           :: axis_data(:)    !< Array of coordinate values
+    CHARACTER(len=*),   INTENT(in)           :: units           !< Units for the axis
+    CHARACTER(len=*),   INTENT(in)           :: cart_name       !< Cartesian axis ("X", "Y", "Z", "T", "U", "N")
```
The `len` for `cart_name` should always be 1.  Is there any reason to not have it be 1?

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+  !> @brief Initialize the axis
+  subroutine diag_axis_init(obj, name, axis_data, units, cart_name, long_name, direction,&
+  & set_name, edges, Domain, Domain2, DomainU, aux, req, tile_count, domain_position )
+    class(diag_axis_t)                       :: obj             !< Diag_axis obj
```
Should this be `intent(out)`?

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+  end subroutine diag_axis_init
+
+  !> @brief Get the length of the axis
+  !> @return axis length
```
Imagine it like this:
```fortran
integer :: axis_length !< axis length
```
Would this be acceptable?  I mean... I guess it would be correct, but it seems a little too short.  Change it or don't, I don't care either way.  This is just nitpicky

uramirez8707 replied:
less is more

thomas-robinson replied:
smh

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+    class(diag_axis_t), intent(inout) :: obj !< diag_axis obj
+    integer                           :: axis_length
+
+    axis_length = obj%length
+
+    !< If the axis is domain decomposed axis_length will be set to the length for the current PE:
+    if (allocated(obj%axis_domain)) axis_length = obj%axis_domain%length(obj%cart_name, obj%domain_position)
```
Should this be an if/else statement?

uramirez8707 replied:
`axis_length` is already defined in L199 if it is not.

thomas-robinson replied:
Yeah, so if `(allocated(obj%axis_domain))` you are setting `axis_length` twice instead of once.
```fortran
if (.not.allocated(obj%axis_domain)) then
    axis_length = obj%length
else
    axis_length = obj%axis_domain%length(obj%cart_name, obj%domain_position)
endif
```
Where or not it evaluates to true, the code will still check the `if` statement, so you are just adding an extra instruction.

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+  end function
+
+  !> @brief Set the subaxis of the axis obj
+  subroutine set_subaxis(obj, bounds)
```
A routine like this already exist.  Are you planning on copying it over/modifying it for the purpose of filling in this object?

uramirez8707 replied:
yes

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+    class(*),         INTENT(IN)    :: bounds(:) !< bounds of the subaxis
+    logical                         :: exists
+
+    !< TO DO: compare bounds
```
These are some big TODOs.  I think that maybe a list of the TODOs should go into an issue for this module.  There seems to be a lot of work that you have planned.

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+!> @addtogroup fms_diag_axis_object_mod
+!> @{
+module fms_diag_axis_object_mod
+  use mpp_domains_mod, only:  domain1d, domain2d, domainUG, mpp_get_compute_domain, CENTER, &
```
What about `NORTH` and `EAST`?

uramirez8707 replied:
They are not being used... It might be used later, but not right now. 
Center is just used to initialize the %domain_position variable:
https://github.com/NOAA-GFDL/FMS/blob/f8610606f52991c97010d5bb79395af6e3dbad25/diag_manager/fms_diag_axis_object.F90#L173

thomas-robinson replied:
They will definitely be used, but you can add them when needed.

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+  !! This type was created to avoid having to send in "Domain", "Domain2", "DomainUG" as arguments into subroutines
+  !! and instead only 1 class(fms_domain_t) argument can be send
+  !> @ingroup diag_axis_object_mod
+  type fms_domain_t
```
this is a diag manager type, so it should be named `fmsDiagDomain_t` or `diagDomain_t`.  Using this name makes it seem like it comes from `fms_mod`.    The same is true for the types that extend this one.

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+
+  !> @brief Get the length of a 2D domain
+  !> @return Length of the 2D domain
+  function get_length(obj, cart_axis, domain_position) &
+  result (length)
+    class(fms_domain_t), INTENT(INOUT) :: obj       !< diag_axis obj
+    character(len=*),    INTENT(IN)    :: cart_axis !< cart_axis of the axis
+    integer,             INTENT(IN)    :: domain_position !< Domain position (CENTER, NORTH, EAST)
```
If you are inquiring about the domain, shouldn't you already know the `cart_axis` in object you are calling it for?  I guess what I'm saying is shouldn't this function be called using the axis object instead of the domain object?  Then you wouldn't need any input arguments other than the object.  Is there a use case for this function where an axis object wouldn't/couldn't be used?

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+  end subroutine set_axis_domain
+
+  !!!!!!!!!!!!!!!!! OTHER FUNCTIONS/SUBROUTINES !!!!!!!!!!!!!!!!!
+  !> @brief Copy a allocatable string
```
Copy *an* allocatable string

thomas-robinson commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+
+  !!!!!!!!!!!!!!!!! OTHER FUNCTIONS/SUBROUTINES !!!!!!!!!!!!!!!!!
+  !> @brief Copy a allocatable string
+  !> @return Allocated string
+  !> TO DO Move this somewhere else?
+  function diag_copy_string(string_in) &
+  result(string_out)
+    character(len=*),             INTENT(IN) :: string_in !< String to copy
+    character(len=:), ALLOCATABLE            :: string_out
+
+    allocate(character(len=len_trim(string_in)) :: string_out)
+    string_out = trim(string_in)
+
+  end function
```
1. Why do you need this at all?  Why not just allocate and copy the string in line?
2. There is a whole module of string routines external to diag manager where this would fit better.

uramirez8707 replied:
it is cleaner and less error prone to do:
```F90
string_out = diag_copy_string(string_in)
```

than

```F90
    allocate(character(len=len_trim(string_in)) :: string_out)
    string_out = trim(string_in)
```

especially when it is done multiple times. 

Moving this function is already in the TO DO, it has to be done through main first ...

thomas-robinson replied:
```fortran
string_out = trim(string_in)
```
You shouldn't need to allocate the string length, the assignment should handle that for you.  Then you don't need a function (which should be named `fms_string_copy`) or an allocate statement.  
https://fortranwiki.org/fortran/files/character_handling_in_Fortran.html

thomas-robinson reviewed (changes requested):
In addition to the comments made in the review:
1. Your type names don't follow the [capitalWordFormat](https://github.com/NOAA-GFDL/FMS/blob/main/CODE_STYLE.md#derived-types) and need to be fixed.
2. You have a lot of big TODOs in this one.  I think that you should make an issue that has the TODOs in it for easier tracking.

uramirez8707 commented on diag_manager/fms_diag_axis_object.F90:
```diff
@@ -0,0 +1,287 @@
+  !> @}
+
+  !> @brief Type to hold the domain info for an axis
+  !! This type was created to avoid having to send in "Domain", "Domain2", "DomainUG" as arguments into subroutines
```
The diag_obj is going to use the diag_axis to figure out what domain the field is in.

uramirez8707 commented:
@thomas-robinson 
I think i addressed all your comments in https://github.com/NOAA-GFDL/FMS/pull/966/commits/eb29e721c954b6ec694c714804cea79660eb14cc or by replying to them. 

I also opened issue #976

thomas-robinson reviewed (commented):
I resolved most of the conversations (some begrudgingly).  There are still two outstanding though that should be addressed.
